libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,354 @@
+				jsonRelated["location"] = toJson(secondaryLocation);
+				jsonDiag["relatedInformation"].append(jsonRelated);
+			}
+
+		diagnosticsBySourceUnit[*location->sourceName].append(jsonDiag);
+	}
+
+	for (string const& sourceUnitName: m_fileReader.sourceCodes() | ranges::views::keys)
+	{
+		Json::Value params;
+		params["uri"] = sourceUnitNameToClientPath(sourceUnitName);
+		params["diagnostics"] = move(diagnosticsBySourceUnit.at(sourceUnitName));
+		m_client.notify("textDocument/publishDiagnostics", move(params));
+	}
+}
+
+bool LanguageServer::run()
+{
+	while (!m_exitRequested && !m_client.closed())
+	{
+		optional<Json::Value> const jsonMessage = m_client.receive();
+		if (!jsonMessage)
+		{
+			m_client.error({}, ErrorCode::ParseError, "Error parsing JSONRPC request.");
+			continue;
+		}
+
+		try
+		{
+			string const methodName = (*jsonMessage)["method"].asString();
+			MessageID const id = (*jsonMessage)["id"];
+
+			if (auto handler = valueOrDefault(m_handlers, methodName))
+				handler(id, (*jsonMessage)["params"]);
+			else
+				m_client.error(id, ErrorCode::MethodNotFound, "Unknown method " + methodName);
+		}
+		catch (exception const& e)
+		{
+			m_client.error({}, ErrorCode::InternalError, "Unhandled exception: "s + e.what());
+		}
+	}
+	return m_shutdownRequested;
+}
+
+void LanguageServer::handleInitialize(MessageID _id, Json::Value const& _args)
+{
+	// The default of FileReader is to use `.`, but the path from where the LSP was started
+	// should not matter.
+	string rootPath("/");
+	if (Json::Value uri = _args["rootUri"])
+	{
+		rootPath = uri.asString();
+		if (!boost::starts_with(rootPath, "file://"))
+		{
+			m_client.error(_id, ErrorCode::InvalidParams, "rootUri only supports file URI scheme.");
+			return;
+		}
+		rootPath = rootPath.substr(7);
+	}
+	else if (Json::Value rootPath = _args["rootPath"])
+		rootPath = rootPath.asString();
+
+	m_fileReader.setBasePath(boost::filesystem::path(rootPath));
+	if (_args["initializationOptions"].isObject())
+		changeConfiguration(_args["initializationOptions"]);
+
+	Json::Value replyArgs;
+	replyArgs["serverInfo"]["name"] = "solc";
+	replyArgs["serverInfo"]["version"] = string(VersionNumber);
+	replyArgs["capabilities"]["textDocumentSync"]["openClose"] = true;
+	replyArgs["capabilities"]["textDocumentSync"]["change"] = 2; // 0=none, 1=full, 2=incremental
+
+	m_client.reply(_id, move(replyArgs));
+}
+
+void LanguageServer::handleWorkspaceDidChangeConfiguration(MessageID, Json::Value const& _args)
+{
+	if (_args["settings"].isObject())
+		changeConfiguration(_args["settings"]);
+}
+
+void LanguageServer::handleTextDocumentDidOpen(MessageID _id, Json::Value const& _args)
+{
+	if (!_args["textDocument"])
+		m_client.error(_id, ErrorCode::RequestFailed, "Text document parameter missing.");
+
+	string text = _args["textDocument"]["text"].asString();
+	string uri = _args["textDocument"]["uri"].asString();
+	m_fileReader.setSourceDirectly(clientPathToSourceUnitName(uri), move(text));
```

Comment:
I wanted to have the explicit call to `clientPathToSourceUnitName` because I wanted to guard against the magic that FileReader does. For example, FileReader always has code paths that use the cwd.